design/design-revised.md

```diff
@@ -0,0 +1,63 @@
+Teleport coding challenge design document by Jin Huang
+
+I) Architecture:
+
+The library will consist of the worker and worker service. At a high level view, a worker will be able
+to perform start, stop, and query status for a command, and the worker service provide a map 
+for the client to have control over multiple workers. The worker service also provides the workers a
+custom logger, which will be used to log output of a process (sdtout and stderr) during execution of a 
+running job to a file or a io.Writer.
+
+The structure will be as follows:
+
+Worker:
+- struct: UID, command, status, error, *logger, mutex
+- functions: NewWorker(*logger, command), Start(), Stop(), Status(), execute(),
+statusUpdate(new status), ExitStatus()
+
+Worker Service:
+- struct: workersmap, *logger, mutex
+- functions: NewWorkerService(), GetWorker(UID), AddWorker(worker)
+
+II) Functions design:
+
+- Error handling will be done approprietly in each function.
+- Locks with sync.mutex will be added where applicable
+
+Worker:
+- NewWorker(): take in a command and a *logger and creates a new worker
+- Start(): Set status of worker as "running", and call Execute(). When Execute() returns finish or 
+error, set status of worker accordingly.
+- Stop(): Set status of a running worker as "killed" and terminates the process
+- Status(): returns the status of a worker (with a lock)
+- execute(): directly execute a command for a process, then pipe both Stdout and Stderr of the process
```

Comment:
> When the process finishes or errors, return back to Start()

Sorry, I'm not sure I follow what this means?

Reply:
execute() will serve as a helper function for Start. It will wait for the job to finish or error and return its exit code; based on this returned code, Start will then update the status of the worker to "finished" or "error". I will update the document the make it more clear.